File: UnitTest/Zeze/Misc/TestServiceManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Zeze.Services.ServiceManager;
using Zeze.Services;
using Zeze.Net;

namespace UnitTest.Zeze.Misc
{
    [TestClass]
    public class TestServiceManager
    {
        [TestInitialize]
        public void TestInit()
        {
            demo.App.Instance.Start();
        }
        [TestCleanup]
        public void TestCleanup()
        {
            demo.App.Instance.Stop();
        }
        TaskCompletionSource<int> future;

        [TestMethod]
        public void TestBase()
        {
            var infos = new ServiceInfos("TestBase");
            infos.Insert(new ServiceInfo("TestBase", "1", 0));
            infos.Insert(new ServiceInfo("TestBase", "3", 0));
            infos.Insert(new ServiceInfo("TestBase", "2", 0));
            Assert.AreEqual("TestBase=[1,2,3,]", infos.ToString());
        }

        [TestMethod]
        public async Task Test1()
        {
            string ip = "127.0.0.1";
            int port = 5001; // c# 版本的Server已经删除，所以不能动态自建服务器测试，使用java版本的，需要先运行准备好。

            System.Net.IPAddress address =
                string.IsNullOrEmpty(ip)
                ? System.Net.IPAddress.Any
                : System.Net.IPAddress.Parse(ip);

            // 后面需要手动销毁重建进行重连测试。不用using了，使用TestCleanup关闭最后的实例。
            var serviceName = "TestServiceManager";

            future = new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
            // for reconnect
            var clientConfig = demo.App.Instance.Zeze.Config;
            var agentConfig = new ServiceConf();
            var agentName = "Zeze.Services.ServiceManager.Agent.Test";
            clientConfig.ServiceConfMap.TryAdd(agentName, agentConfig);
            agentConfig.AddConnector(new Connector(ip, port));
            using var agent = new Agent(demo.App.Instance.Zeze, agentName);
            agent.Client.Start();
            await agent.RegisterService(serviceName, "1", 0, "127.0.0.1", 1234);
            agent.OnChanged = (state) =>
            {
                Console.WriteLine("OnChanged: " + state);
                this.future.SetResult(0);
            };
            await agent.SubscribeService(serviceName);
            var load = new ServerLoad()
            {
                Ip = "127.0.0.1",
                Port = 1234,
            };
            agent.SetLoad(load);
            Console.WriteLine("ConnectNow");
            future.Task.Wait();

            Console.WriteLine("RegisterService 2");
            future = new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
            await agent.RegisterService(serviceName, "2", 0);
            future.Task.Wait();

            // 改变订阅类型
            Console.WriteLine("Change Subscribe type");
            await agent.UnSubscribeService(serviceName);
            future = new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
            await agent.SubscribeService(serviceName);
            future.Task.Wait();

            agent.SubscribeStates.TryGetValue(serviceName, out var state);
            object anyState = this;
            state.SetIdentityLocalState("1", anyState);
            state.SetIdentityLocalState("2", anyState);
            state.SetIdentityLocalState("3", anyState);

            Console.WriteLine("RegisterService 3");
            future = new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
            await agent.RegisterService(serviceName, "3", 0);
            future.Task.Wait();
        }
    }
}

File: ZezeJava/ZezeJavaTest/src/UnitTest/Zeze/Misc/TestServiceManager.java
package UnitTest.Zeze.Misc;

import Zeze.Services.ServiceManager.BServerLoad;
import Zeze.Services.ServiceManager.BServiceInfo;
import Zeze.Services.ServiceManager.BServiceInfos;
import Zeze.Services.ServiceManager.BSubscribeInfo;
import Zeze.Util.TaskCompletionSource;
import demo.App;
import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

public class TestServiceManager {
	@Test
	public void testServiceInfos() {
		var infos = new BServiceInfos();
		infos.insert(new BServiceInfo("TestBase", "1", 0));
		infos.insert(new BServiceInfo("TestBase", "3", 0));
		infos.insert(new BServiceInfo("TestBase", "2", 0));
		var it = infos.getSortedIdentities().iterator();
		Assert.assertTrue(it.hasNext());
		Assert.assertEquals("1", it.next().getServiceIdentity());
		Assert.assertTrue(it.hasNext());
		Assert.assertEquals("2", it.next().getServiceIdentity());
		Assert.assertTrue(it.hasNext());
		Assert.assertEquals("3", it.next().getServiceIdentity());
		Assert.assertFalse(it.hasNext());
	}

	@Before
	public void before() throws Exception {
		App.Instance.Start();
	}

	@After
	public void after() throws Exception {
		//App.Instance.Stop();
	}

	TaskCompletionSource<Integer> future;

	@Test
	public void test1() {
		if (null == App.Instance.Zeze.getServiceManager())
			return; // disable

		var serviceName = "TestServiceManager";
		future = new TaskCompletionSource<>();

		var agent = App.Instance.Zeze.getServiceManager();
		agent.registerService(new BServiceInfo(serviceName, "1", 0, "127.0.0.1", 1234));
		agent.setOnChanged((state) -> {
			System.out.println("OnChanged 1:" + state);
			this.future.setResult(0);
		});
		agent.setOnSetServerLoad((load) -> {
			System.out.println("OnSetLoad " + load);
			this.future.setResult(0);
		});
		agent.subscribeService(new BSubscribeInfo(serviceName));
		var load = new BServerLoad();
		load.ip = "127.0.0.1";
		load.port = 1234;
		System.out.println("WaitOnSetLoad");
		agent.setServerLoad(load);
		future.await();

		System.out.println("RegisterService 2");
		future = new TaskCompletionSource<>();
		System.out.println("WaitOnChanged 2");
		agent.registerService(new BServiceInfo(serviceName, "2"));
		future.await();

		var state = agent.getSubscribeStates().get(serviceName);
		Object anyState = this;
		state.setIdentityLocalState("1", anyState);
		state.setIdentityLocalState("2", anyState);
		state.setIdentityLocalState("3", anyState);

		System.out.println("RegisterService 3");
		future = new TaskCompletionSource<>();
		System.out.println("WaitOnChanged 3");
		agent.registerService(new BServiceInfo(serviceName, "3"));
		future.await();
	}
}
